MaxText/layers/llama4.py

```diff
@@ -0,0 +1,222 @@
+    lnx_rms = models.RMSNorm(
+        dtype=cfg.dtype,
+        weight_dtype=cfg.weight_dtype,
+        name="pre_self_attention_layer_norm",
+        kernel_axes=("norm",),
+        epsilon=cfg.normalization_layer_epsilon,
+    )
+    lnx = lnx_rms(inputs)
+
+    lnx = nn.with_logical_constraint(lnx, ("activation_batch", "activation_norm_length", "activation_embed"))
+    # Instead of scaling the query values in the checkpoint conversion (`llama_or_mistral_ckpt`)
+    # we'll do it dynamically in the forward pass of Attention
+    query_pre_attn_scalar = cfg.head_dim**-0.5
+
+    # Self-attention block
+    attention_layer = Attention(
+        config=cfg,
+        num_query_heads=cfg.num_query_heads,
+        num_kv_heads=cfg.num_kv_heads,
+        head_dim=cfg.head_dim,
+        max_target_length=cfg.max_target_length,
+        max_prefill_predict_length=cfg.max_prefill_predict_length,
+        attention_kernel=cfg.attention,
+        mesh=mesh,
+        dtype=cfg.dtype,
+        weight_dtype=cfg.weight_dtype,
+        dropout_rate=cfg.dropout_rate,
+        name="self_attention",
+        float32_qk_product=cfg.float32_qk_product,
+        float32_logits=cfg.float32_logits,
+        quant=self.quant,
+        kv_quant=quantizations.configure_kv_quant(cfg),
+        prefill_cache_axis_order=tuple([int(i) for i in cfg.prefill_cache_axis_order.split(",")]),
+        ar_cache_axis_order=tuple([int(i) for i in cfg.ar_cache_axis_order.split(",")]),
+        compute_axis_order=tuple([int(i) for i in cfg.compute_axis_order.split(",")]),
+        reshape_q=cfg.reshape_q,
+        use_ragged_attention=cfg.use_ragged_attention,
+        ragged_block_size=cfg.ragged_block_size,
+        is_nope_layer=self.is_nope_layer,
+        use_qk_norm=cfg.use_qk_norm,
+        query_pre_attn_scalar=query_pre_attn_scalar,
+    )
+
+    attention_lnx = attention_layer(
+        lnx,
+        lnx,
+        decoder_positions,
+        decoder_segment_ids=decoder_segment_ids,
+        deterministic=deterministic,
+        model_mode=model_mode,
+        slot=slot,
+        page_state=page_state,
+        previous_chunk=previous_chunk,
+    )
+
+    attention_lnx = nn.with_logical_constraint(
+        attention_lnx, ("activation_batch", "activation_norm_length", "activation_embed")
+    )
+    intermediate_inputs = inputs + attention_lnx
+
+    # Fully Connected
+    hidden_states = models.RMSNorm(
+        dtype=cfg.dtype,
+        weight_dtype=cfg.weight_dtype,
+        name="post_self_attention_layer_norm",
+        kernel_axes=("norm",),
+        epsilon=cfg.normalization_layer_epsilon,
+    )(intermediate_inputs)
+    hidden_states = nn.with_logical_constraint(
+        hidden_states, ("activation_batch", "activation_norm_length", "activation_embed")
+    )
+
+    load_balance_loss = None
+    assert (
+        cfg.num_experts > 1
+    ), "Expected the Llama4 config to have num_experts > 1.  Note that MaxText only supports Llama4 Scout (MoE-only) for now!"
```

Comment:
nit: will it be better we move `Note that MaxText only supports Llama4 Scout (MoE-only) for now!` to head of file near `The Decoder Layer specific for LLama4`?